execution/shuffle.go

```diff
@@ -0,0 +1,449 @@
+		}
+
+		outRecordStreams[partition] = rs
+		outExecOutputs[partition] = execOutput
+	}
+
+	// Now iteratively start all shuffles.
+	for len(nextShuffles) > 0 {
+		newNextShuffles := make(map[string]ShuffleData)
+
+		for _, shuffle := range nextShuffles {
+			tmpNextShuffles, err := shuffle.Shuffle.StartSources(ctx, stateStorage, tx, shuffle.ShuffleID, shuffle.Variables)
+			if err != nil {
+				return nil, nil, errors.Wrapf(err, "couldn't start sources for shuffle with id %v", err)
+			}
+
+			// Deduplicate shuffles by their ID's
+			for id, data := range tmpNextShuffles {
+				newNextShuffles[id] = data
+			}
+		}
+
+		nextShuffles = newNextShuffles
+	}
+
+	return outRecordStreams, outExecOutputs, nil
+}
+
+type pipelineMetadataContextKey struct{}
+
+type PipelineMetadata struct {
+	// The ID for the next shuffle.
+	NextShuffleID *ShuffleID
+
+	// The partition of the current stream.
+	Partition int
+}
+
+func (id *ShuffleID) AsPrefix() []byte {
+	return []byte("$" + id.Id + "$")
+}
+
+func (id *ShuffleID) AsMapKey() string {
+	return id.Id
+}
+
+type Shuffle struct {
+	outputPartitionCount int
+	strategyPrototype    ShuffleStrategyPrototype
+	sources              []Node
+}
+
+func NewShuffle(outputPartitionCount int, strategyPrototype ShuffleStrategyPrototype, sources []Node) *Shuffle {
+	return &Shuffle{
+		outputPartitionCount: outputPartitionCount,
+		strategyPrototype:    strategyPrototype,
+		sources:              sources,
+	}
+}
+
+func (s *Shuffle) Get(ctx context.Context, variables octosql.Variables, streamID *StreamID) (RecordStream, *ExecutionOutput, error) {
+	pipelineMetadata := ctx.Value(pipelineMetadataContextKey{}).(PipelineMetadata)
+
+	receiver := NewShuffleReceiver(streamID, pipelineMetadata.NextShuffleID, len(s.sources), pipelineMetadata.Partition)
+	execOutput := NewExecutionOutput(receiver, map[string]ShuffleData{
+		pipelineMetadata.NextShuffleID.AsMapKey(): {
+			Shuffle:   s,
+			Variables: variables,
+			ShuffleID: pipelineMetadata.NextShuffleID,
+		},
+	})
+
+	return receiver, execOutput, nil
+}
+
+func (s *Shuffle) StartSources(ctx context.Context, stateStorage storage.Storage, tx storage.StateTransaction, shuffleID *ShuffleID, variables octosql.Variables) (map[string]ShuffleData, error) {
+	// Create an ID for the possible _even next_ shuffle.
+	nextShuffleIDRaw, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString("next_shuffle_id"))
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get next shuffle id")
+	}
+	nextShuffleID := (*ShuffleID)(nextShuffleIDRaw)
+
+	nextShuffles := make(map[string]ShuffleData)
+
+	for partition, node := range s.sources {
+		sourceStreamID, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString(fmt.Sprintf("shuffle_input_%d", partition)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream ID for shuffle with ID %s", shuffleID.Id)
+		}
+
+		newPipelineMetadata := PipelineMetadata{
```

Comment:
I don't really get what this is for

Reply:
I think I answered in a comment above, let me know if not.